DESIGN_GUIDELINES.md

```diff
@@ -0,0 +1,671 @@
+:x:                     | Strip quotes                                      | Should be handled by the shell
+:x:                     | Cast strings to lowercase / uppercase             | Avoid string value manipulation if not needed.
+
+
+
+### Commands and subcommands
+
+#### Standardized "CRUD" commands and aliases
+
+- `create`
+- `remove`, `rm`
+- `update`
+- `list`, `ls`
+- `inspect` (JSON)
+
+Command                          | Aliases     | Description
+---------------------------------|-------------|--------------------------------------------------
+`docker <object> create`         | -           | Creates a new `<object>`
+`docker <object> list`           | `ls` / `ps` | Presents a list of `<object>` (table view by default)
+`docker <object> inspect <id>`   | -           | Provides low-level information about `<object> <id>` in JSON format
+`docker <object> update <id>`    | -           | Updates `<object> <id>`
+`docker <object> remove <id>`    | `rm`        |
+`docker <object> prune`          | -           |
+
+
+
+
+### Flags
+
+#### Shorthand (single-letter) flags
+
+
+
+#### Standardized flags
+
+Flag              | Generally used on   | Description
+------------------|---------------------|----------------------------------------------------------------
+`--format` / `-f` | `list`, `inspect`   | Pretty-print objects using a Go template
+`--filter` / `-f` | `list`              | Filter objects based on conditions provided
+`--no-trunc`      | `list`              | List outputs can truncate columns to save screen-space. The `--no-trunc` option prints columns without truncating
+`--quiet` / `-q`  | `list`              | For list outputs; only print object ID's.
+
+
+
+
+#### File flags
+
+- absolute vs relative paths
+- support for stdin (CONVENTION??)
+
+### Feedback on successful operations
+
+Successful operations on an object should print the object's identifier on `stdout`
+on success; doing so enables users to consume the output for scripting.
+
+For example, the following command creates two volumes, using different drivers,
+and attaches those volumes to a new container:
+
+```bash
+docker container run \
+  --volume $(docker volume create --driver=foo):/somewhere \
+  --volume $(docker volume create --driver=bar):/somewhere-else \
+  busybox
+```
+
+
+Either `ID` or `name` are acceptable, as long as the identifier can be used
+to reference the object.
+
+Some examples of commands that follow this design:
+
+
+Creating a container prints the container's `ID` on `stdout`
+
+```bash
+docker container create --name test busybox
+cc1555ede50a11f02a3ef6cc8eedd9f78f6299a2f2b7efdc8a91fbefb8fc194a
+```
+
+
+Removing a container prints the reference that was given
+
+```bash
+docker container rm test
+test
+
+docker container rm 057d35243903e522c70ae2dfab5f706e4ea3cc1ae4c38cce955a99be3d09cfd1
+057d35243903e522c70ae2dfab5f706e4ea3cc1ae4c38cce955a99be3d09cfd1
+```
+
+
+
+### Use of stdout and stderr
```

Comment:
What about adding a paragraph on logging/verbose/debug? As far as I can remember it's quite a mess, some commands add verbosity, but most don't. This paragraph could then at least explain the situation and what we want?

Reply:
Yes, good idea to explain a bit more